### PR TITLE
Fix commutation of classical controls

### LINKED ISSUE #4870: `ClassicallyControlledOperation` breaks assumption that disjoint operations commute.
### Description of the issue
`ClassicallyControlledOperation` was recently introduced to enable classical controls in Cirq. The method takes as input a `sub_operation` and a list of `conditions`, which correspond to measurement keys of previous measurements. 

https://github.com/quantumlib/Cirq/blob/6119620178d7d1d0f11c5f2b73998fb2b66fcd2d/cirq-core/cirq/ops/classically_controlled_operation.py#L49-L53

The operations is implemented in a way that the qubit set it acts on remains the same as the set of qubits of the underlying operation to be controlled, and the operation "assumes" that measurements corresponding to the conditions have already occurred to the left of the operation. 

**This is a big problem for transformers because this introduces an implicit dependency between a measurement operation and a classically controlled operation which makes use of these measurement results.** 

### How to reproduce the issue

```python
>>> import cirq
>>> q = cirq.LineQubit.range(2)
>>> c = cirq.Circuit(cirq.H(q[0]), cirq.measure(q[0], key='m'), cirq.X(q[1]).with_classical_controls('m'))
>>> print(c)
0: ───H───M───────
          ║
1: ───────╫───X───
          ║   ║
m: ═══════@═══^═══
>>> cirq.SynchronizeTerminalMeasurements().optimize_circuit(c)
>>> print(c)
0: ───H───────────M───
                  ║
1: ───────────X───╫───
              ║   ║
m: ═══════════^═══@═══
>>> cirq.Simulator().simulate(c)
ValueError: Measurement key m missing when testing classical control
```

**Cirq version**
```
0.14.0.dev
```

cc @95-martin-orion @daxfohl

### LINKED ISSUE #4880: Fix `__commutes__` for `ClassicallyControlledOperation`
**Description of the issue**
As discussed in https://github.com/quantumlib/Cirq/issues/4870, `__commutes__` for `CCO` is currently broken and should be fixed. 

Opening a separate issue to track this smaller change. 

**How to reproduce the issue**

```python
>>> import cirq
>>> q = cirq.LineQubit.range(2)
>>> op1, op2 = cirq.measure(q[0], key='m'), cirq.X(q[1]).with_classical_controls('m')
>>> cirq.commutes(op1, op2), cirq.commutes(op2, op1) # Both should return False
(True, True)
```

**Cirq version**
0.14.0.dev

### LINKED ISSUE #4882: Fix `circuit.append()` to use `protocols.commute` to verify if two operations can be "crossed" or not.
**Description of the issue**
Since #4870 breaks the assumption that two operations acting on disjoint qubits always commute, `circuit.append()` should use `protocols.commute` to find the leftmost moment in which an operation can be inserted. 

As part of the work done for `CCO`, this was implemented by adding a special case of checking whether the current operation is a CCO and if yes, whether the previous moment has a measurement with matching control keys. 

https://github.com/quantumlib/Cirq/blob/90679d27e0967c0a7fe59848e789a5d6ba49b66d/cirq-core/cirq/circuits/circuit.py#L1972-L1975

This is problematic because it works only in one direction and breaks optimizers which relies on doing operations on the circuit in reverse order. 

**How to reproduce the issue**

```python
>>> import cirq
>>> q =  cirq.LineQubit.range(2)
>>> c = cirq.Circuit(cirq.Moment(cirq.X(q[0])), cirq.Moment(cirq.measure(q[1], key="m")), cirq.Moment(cirq.X(q[0]).with_classical_controls("m")))
>>> print(c)
0: ───X───────X───
              ║
1: ───────M───╫───
          ║   ║
m: ═══════@═══^═══
>>> cirq.AlignLeft().optimize_circuit(c) # Works as expected.
0: ───X───X───
          ║
1: ───M───╫───
      ║   ║
m: ═══@═══^═══
>>> cirq.AlignRight().optimize_circuit(c) # Does not work as expected because implementation depends on inserting operations in the reverse circuit.
>>> print(c) # Wrong output
          ┌──┐
0: ───X────X─────
           ║
1: ────────╫M────
           ║║
m: ════════^@════
          └──┘

```

**Cirq version**
0.14.0.dev

### PR DESCRIPTION
**Pull request by daxfohl**
Fixes #4870: Adds logic to ensure classically controlled ops do not commute with measurements of the same key (fixes #4880). Also hardens the logic in circuit.append to apply this logic bidirectionally (fixes #4882 too).

### CODE REVIEW

**Comment by MichaelBroughton** [comment]
Does this also fix issues with `optimizers` and `decompose` ?

**Comment by daxfohl** [comment]
I can't make a blanket statement, but most optimizers should either use commutes protocol or circuit.append, and those would be fixed.

I'm not aware of an issue with decompose.

**Comment by MichaelBroughton** [comment]
Running the test snippet from #4870 it looks like the measurements are still hopping over one another and the simulation is breaking. Am I missing something ?
```python3
q = cirq.LineQubit.range(2)
c = cirq.Circuit(cirq.H(q[0]), cirq.measure(q[0], key='m'), cirq.X(q[1]).with_classical_controls('m'))
print(c)

cirq.SynchronizeTerminalMeasurements().optimize_circuit(c)
print(c)
cirq.Simulator().simulate(c)
```

**Comment by daxfohl** [comment]
Ah, I didn't realize it's an inplace transformer, so the test I added doesn't do anything. Changing my test to `after=circuit.copy()` repros the problem.

Anyway, that whole transformer has major issues. It doesn't handle multi-qubit measurements either. The following breaks:

```python
def test_multi_qubit():
    q0, q1 = cirq.LineQubit.range(2)
    circuit = cirq.Circuit(
        cirq.measure(q0, q1, key='m'), cirq.H(q1)
    )
    assert_optimizes(before=circuit, after=circuit.copy())
```

That said, I made the same error in the alight_right test, and after fixing the test, it turns red again. This time the problem is just that `Moment._control_keys_` is not defined. Pushing a fix for that in a sec.

**Comment by daxfohl** [comment]
I'll add some extra tests to the commutes protocol before merging.

File: cirq-core/cirq/ops/raw_types.py [hunk 1]
```diff
@@ -572,13 +572,22 @@ def _commutes_(
...
+
         if hasattr(other, 'qubits') and set(self.qubits).isdisjoint(other.qubits):
             return True
 
         from cirq import circuits
 
-        circuit12 = circuits.Circuit(self, other)
-        circuit21 = circuits.Circuit(other, self)
+        self_raw = self.without_classical_controls()
+        other_raw = other.without_classical_controls()
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/raw_types.py:588]: Add a brief comment explaining why this is needed.

**Review by tanujkhattar — APPROVED** [review]
LGTM % nit.

File: cirq-core/cirq/optimizers/align_left_test.py [hunk 2]
```diff
@@ -44,3 +44,11 @@ def test_align_left():
...
         ),
     )
+
+
+def test_classical_control():
+    q0, q1 = cirq.LineQubit.range(2)
+    circuit = cirq.Circuit(
+        cirq.H(q0), cirq.measure(q0, key='m'), cirq.X(q1).with_classical_controls('m')
+    )
+    assert_optimizes(before=circuit, after=circuit.copy())
```
Comment (tanujkhattar) [inline cirq-core/cirq/optimizers/align_left_test.py:54]: Can you add these tests to https://github.com/quantumlib/Cirq/blob/master/cirq-core/cirq/transformers/align_test.py instead? `AlignLeft` / `AlignRight` are now deprecated after https://github.com/quantumlib/Cirq/pull/4891

**Pull request merged**